<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>S4 class for an object representing a Genome-wide Assocaition Study. — GWASdata • kangar00</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="S4 class for an object representing a Genome-wide Assocaition Study. — GWASdata"><meta property="og:description" content="S4 class for an object representing a Genome-wide Assocaition Study.
'GWASdata' is a GWASdata object constructor.
show displays basic information on GWASdata object
summary summarizes the content of a GWASdata object 
and gives an overview about the information included in a 
GWASdata object. Summary statistics for phenotype and genotype 
data are calculated.
GeneSNPsize creates a data.frame of pathway
names with numbers of snps and genes in each pathway."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">kangar00</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>S4 class for an object representing a Genome-wide Assocaition Study.</h1>
    
    <div class="hidden name"><code>GWASdata-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>S4 class for an object representing a Genome-wide Assocaition Study.</p>
<p><code>'GWASdata'</code> is a GWASdata object constructor.</p>
<p><code>show</code> displays basic information on <code>GWASdata</code> object</p>
<p><code>summary</code> summarizes the content of a <code>GWASdata</code> object 
and gives an overview about the information included in a 
<code>GWASdata</code> object. Summary statistics for phenotype and genotype 
data are calculated.</p>
<p><code>GeneSNPsize</code> creates a <code>data.frame</code> of <code><a href="pathway-class.html">pathway</a></code>
names with numbers of snps and genes in each <code><a href="pathway-class.html">pathway</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GWASdata</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for ANY</span></span>
<span><span class="fu">GWASdata</span><span class="op">(</span><span class="va">geno</span>, <span class="va">anno</span>, pheno <span class="op">=</span> <span class="cn">NULL</span>, desc <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for GWASdata</span></span>
<span><span class="fu">show</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for GWASdata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for GWASdata</span></span>
<span><span class="fu">GeneSNPsize</span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>A <code>GWASdata</code> object.</p></dd>


<dt>...</dt>
<dd><p>Further arguments can be added to the function.</p></dd>


<dt>geno</dt>
<dd><p>An object of any type, including the genotype information.</p></dd>


<dt>anno</dt>
<dd><p>A <code>data.frame</code> containing the annotation file for the 
<code>GWASdata</code> object.</p></dd>


<dt>pheno</dt>
<dd><p>A <code>data.frame</code> specifying individual IDs, phenotypes and
covariates to be included in the regression model.</p></dd>


<dt>desc</dt>
<dd><p>A <code>character</code> giving the GWAS description, e.g. name of study.</p></dd>

</dl></div>
    <div id="methods-by-generic-">
    <h2>Methods (by generic)</h2>
    
<ul><li><p><code>GeneSNPsize(GWASdata)</code>: creates a <code>data.frame</code> of <code><a href="pathway-class.html">pathway</a></code> names with numbers 
of snps and genes in each pathway.</p></li>
</ul></div>
    <div id="slots">
    <h2>Slots</h2>
    

<dl><dt><code>geno</code></dt>
<dd><p>An object of any type, including genotype information. The format
needs to be one line per individual and on colum per SNP in minor-allele 
coding (0,1,2). Other values between 0 and 2, as from impute dosages, are 
allowed. Missing values must be imputed prior to creation of a <code>GWASdata</code> object.</p></dd>


<dt><code>anno</code></dt>
<dd><p>A <code>data.frame</code> mapping SNPs to genes and genes to
pathways. Needs to include the columns 'pathway' (pathway ID, e.g. hsa
number from KEGG database), 'gene' (gene name (hgnc_symbol)), 'chr'
(chromosome), 'snp' (rsnumber) and 'position' (base pair position of SNP).</p></dd>


<dt><code>pheno</code></dt>
<dd><p>A <code>data.frame</code> specifying individual IDs, phenotypes and
covariates to be included in the regression model e.g. ID, pheno, sex,
pack.years. Note: IDs have to be in the first column!</p></dd>


<dt><code>desc</code></dt>
<dd><p>A <code>character</code> giving the GWAS description, e.g. name of study.</p></dd>


</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>Juliane Manitz, Stefanie Friedrichs</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># create gwas data object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span><span class="st">'GWASdata'</span>, pheno<span class="op">=</span><span class="va">pheno</span>, geno<span class="op">=</span><span class="va">geno</span>, anno<span class="op">=</span><span class="va">anno</span>, desc<span class="op">=</span><span class="st">"some study"</span><span class="op">)</span> </span></span>
<span class="r-in"><span><span class="co"># show and summary methods</span></span></span>
<span class="r-in"><span><span class="va">gwas</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class GWASdata from some study</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phenotypes for 50 individuals: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       pheno sex age</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ind01     0   0  23</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ind02     0   1  39</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ind03     0   0  45</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ind04     0   0  52</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ind05     1   1  20</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ind06     0   1  30</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class GWASdata from some study</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Phenotypes for 50 individuals: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      pheno           sex           age       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Min.   :0.00   Min.   :0.0   Min.   :20.00  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  1st Qu.:0.00   1st Qu.:0.0   1st Qu.:24.25  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Median :1.00   Median :1.0   Median :36.00  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Mean   :0.52   Mean   :0.6   Mean   :35.82  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  3rd Qu.:1.00   3rd Qu.:1.0   3rd Qu.:45.00  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Max.   :1.00   Max.   :1.0   Max.   :55.00  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genotypes: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of genes and SNPs per pathway:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          genes SNPs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> hsa04020   180 1335</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> hsa04022   167 1264</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> hsa04024   200 1457</span>
<span class="r-in"><span><span class="co"># SNPs and genes in pathway</span></span></span>
<span class="r-in"><span><span class="fu">GeneSNPsize</span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          genes SNPs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> hsa04020   180 1335</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> hsa04022   167 1264</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> hsa04024   200 1457</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Juliane Manitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

