<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>A function to calculate the p-values for kernel matrices. — lkmt_test • kangar00</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="A function to calculate the p-values for kernel matrices. — lkmt_test"><meta property="og:description" content="For parameter 'satt' a pathway's influence on the probability of 
beeing a case is evaluated in the logistic kernel machine test and p-values 
are determined using a Sattherthwaite approximation as described by Dan Schaid.
For parameter 'davies' a pathways influence on the probability
of beeing a case is evaluated using the p-value calculation method described 
by Davies. Here the function davies from package 
CompQuadForm is used."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">kangar00</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A function to calculate the p-values for kernel matrices.</h1>
    
    <div class="hidden name"><code>lkmt_test.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For parameter <code>'satt'</code> a pathway's influence on the probability of 
beeing a case is evaluated in the logistic kernel machine test and p-values 
are determined using a Sattherthwaite approximation as described by Dan Schaid.</p>
<p>For parameter <code>'davies'</code> a pathways influence on the probability
of beeing a case is evaluated using the p-value calculation method described 
by Davies. Here the function <code><a href="https://rdrr.io/pkg/CompQuadForm/man/davies.html" class="external-link">davies</a></code> from package 
<span class="pkg">CompQuadForm</span> is used.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">lkmt_test</span><span class="op">(</span><span class="va">formula</span>, <span class="va">kernel</span>, <span class="va">GWASdata</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"satt"</span>, <span class="st">"davies"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for matrix</span></span>
<span><span class="fu">score_test</span><span class="op">(</span><span class="va">x1</span>, <span class="va">x2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for matrix</span></span>
<span><span class="fu">davies_test</span><span class="op">(</span><span class="va">x1</span>, <span class="va">x2</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>formula</dt>
<dd><p>The formula to be used for the regression nullmodel.</p></dd>


<dt>kernel</dt>
<dd><p>An object of class <code>kernel</code> including the pathway 
representing kernel-matrix based on which the test statistic will be calculated.</p></dd>


<dt>GWASdata</dt>
<dd><p>A <code>GWASdata</code> object stating the data used in analysis.</p></dd>


<dt>method</dt>
<dd><p>A <code>character</code> specifying which method will be used for 
p-value calculation. Available are <code>'satt'</code> for the Satterthwaite 
approximation and <code>'davies'</code> for Davies' algorithm. For more details 
see the references.</p></dd>


<dt>...</dt>
<dd><p>Further arguments can be given to the function.</p></dd>


<dt>x1</dt>
<dd><p>A <code><a href="make_psd-matrix-method.html">matrix</a></code> which is the
similarity matrix calculated for the pathway to be tested.</p></dd>


<dt>x2</dt>
<dd><p>An <code>lm</code> or <code>glm</code> object of the nullmodel with fixed 
effects covariates included, but no genetic random effects.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An <code>lkmt</code> object including the following test results</p><ul><li><p>The formula of the regression nullmodel used in the variance component test.</p></li>
<li><p>An object of class <code><a href="kernel-class.html">kernel</a></code> including the similarity matrix of the individuals based on which the pathways influence is evaluated.</p></li>
<li><p>An object of class <code><a href="GWASdata-class.html">GWASdata</a></code> stating the data on which the test was conducted.</p></li>
<li><p>statistic A <code>vector</code> giving the value of the variance component test statistic.</p></li>
<li><p>df A <code>vector</code> giving the number of degrees of freedom.</p></li>
<li><p>p.value A <code>vector</code> giving the p-value calculated for the pathway in the variance component test.</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>For details on the variance component test</p><ul><li><p>Wu MC, Kraft P, Epstein MP, Taylor DM, Chanock SJ, Hunter DJ, Lin X: Powerful SNP-Set Analysis for Case-Control Genome-Wide Association Studies. Am J Hum Genet 2010, 86:929-42</p></li>
<li><p>Liu D, Lin X, Ghosh D: Semiparametric regression of multidimensional genetic pathway data: least-squares kernel machines and linear mixed models. Biometrics 2007, 63(4):1079-88.</p></li>
</ul><p>For details on the p-value calculation see</p><ul><li><p>Schaid DJ: Genomic Similarity and Kernel Methods I: Advancements by Building on Mathematical and Statistical Foundations. Hum Hered 2010, 70:109-31</p></li>
</ul><ul><li><p>Davies R: Algorithm as 155: the distribution of a linear combination of chi-2 random variables. J R Stat Soc Ser C 1980, 29:323-333.</p></li>
</ul></div>
    <div id="author">
    <h2>Author</h2>
    <p>Stefanie Friedrichs, Juliane Manitz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hsa04020</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">net_kernel</span> <span class="op">&lt;-</span> <span class="fu"><a href="calc_kernel.html">calc_kernel</a></span><span class="op">(</span><span class="va">gwas</span>, <span class="va">hsa04020</span>, knots<span class="op">=</span><span class="cn">NULL</span>, type<span class="op">=</span><span class="st">'net'</span>, calculation<span class="op">=</span><span class="st">'cpu'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">lkmt_test</span><span class="op">(</span><span class="va">pheno</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">age</span>, <span class="va">net_kernel</span>, <span class="va">gwas</span>, method<span class="op">=</span><span class="st">'satt'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class lkmt:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GWAS data: example study </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Kernel: network </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Pathway: hsa04020 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The score test results in the p-value:  0.3711717 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Juliane Manitz.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

